src/nvim/event/socket.c

```diff
@@ -43,16 +43,21 @@ int socket_watcher_init(Loop *loop, SocketWatcher *watcher,
     *host_end = '\0';
     uv_getaddrinfo_t request;
 
-    int retval = uv_getaddrinfo(&loop->uv, &request, NULL, addr, port,
-                                &(struct addrinfo){
-                                  .ai_family = AF_UNSPEC,
-                                  .ai_socktype = SOCK_STREAM,
-                                });
-    *host_end = ':';
-    if (retval != 0) {
+getaddrinfo:
+    ret = uv_getaddrinfo(&loop->uv, &request, NULL, addr, port,
+                         &(struct addrinfo){
+                           .ai_family = AF_UNSPEC,
+                           .ai_socktype = SOCK_STREAM,
+                         });
+    if (ret != 0) {
       ELOG("Host lookup failed: %s", addr);
-      return retval;
+      if (strcmp(addr, "localhost") == 0) {
+        xstrlcpy(addr, "127.0.0.1", sizeof(*addr));
```

Comment:
`sizeof(*addr)` is 1, the size of a `char`.  You would need `strlen() + 1`, but I think we should just remove the special casing for localhost.  If a user can't specify a valid hostname, then they should get an error.